### ISSUE TITLE
Boolean expression a != c && (b || (c || a)) and a != c && (b || c || a) lead to different (unoptimal) code

### ISSUE THREAD

**Issue by HendrikHuebner** [llvm/llvm-project#95946]
The following two functions lead to different results:

```c++
bool fn1(bool a, bool b, bool c) {
    return a != c && (b || (c || a));
}

bool fn2(bool a, bool b, bool c) {
    return a != c && (b || c || a);
}

```
https://godbolt.org/z/qMe1sMh6M

This expression is equivalent to `a != c`, which seems it seems to be simplified to nicely in fn2 but not fn1. I think this is a codegen issue, since the expressions are semantically equivalent but generate different IR, since fn1 is generated with an extra block/branch.

**Comment by wangbyby** [comment]
The IR from start is different. 
fn1 has an extra block due to `(c || a)`. 
src ir is
```llvmir
lor.end:                                          ; preds = %lor.rhs7, %lor.rhs
  %5 = phi i1 [ true, %lor.rhs ], [ %tobool8, %lor.rhs7 ]
  br label %lor.end9, !dbg !37

lor.end9:                                         ; preds = %lor.end, %land.rhs
  %6 = phi i1 [ true, %land.rhs ], [ %5, %lor.end ]
  br label %land.end
```
And target ir is
```ir
lor.end9:                                         ; preds = %lor.end, %land.rhs
  %6 = phi i1 [ true, %land.rhs ], [ true, %lor.rhs ], [ %tobool8, %lor.rhs7 ]
  br label %land.end
```
Maybe do this in `simplifycfg` is fine?
https://godbolt.org/z/6ohPTsGx3

**Comment by HendrikHuebner** [comment]
Yes, I think you are right. At first I though it would make sense to have clangs codegen try to generate fewer blocks from boolean expressions. For example, by trying out different orderings for the or-ed booleans. This seems a little bit more complicated though.

**Comment by wangbyby** [comment]
And I found code in `llvm\lib\Transforms\Utils\SimplifyCFG.cpp:FoldTwoEntryPHINode` would create the select inst:
```cpp
  while (PHINode *PN = dyn_cast<PHINode>(BB->begin())) {
    if (isa<FPMathOperator>(PN))
      Builder.setFastMathFlags(PN->getFastMathFlags());

    // Change the PHI node into a select instruction.
    Value *TrueVal = PN->getIncomingValueForBlock(IfTrue);
    Value *FalseVal = PN->getIncomingValueForBlock(IfFalse);

    Value *Sel = Builder.CreateSelect(IfCond, TrueVal, FalseVal, "", DomBI);
    PN->replaceAllUsesWith(Sel);
    Sel->takeName(PN);
    PN->eraseFromParent();
  }
```
And I didn't found where to fold basicblocks with phinode.

**Comment by wangbyby** [comment]
> The IR from start is different. fn1 has an extra block due to `(c || a)`. src ir is
> 
> ```
> lor.end:                                          ; preds = %lor.rhs7, %lor.rhs
>   %5 = phi i1 [ true, %lor.rhs ], [ %tobool8, %lor.rhs7 ]
>   br label %lor.end9, !dbg !37
> 
> lor.end9:                                         ; preds = %lor.end, %land.rhs
>   %6 = phi i1 [ true, %land.rhs ], [ %5, %lor.end ]
>   br label %land.end
> ```
> 
> And target ir is
> 
> ```
> lor.end9:                                         ; preds = %lor.end, %land.rhs
>   %6 = phi i1 [ true, %land.rhs ], [ true, %lor.rhs ], [ %tobool8, %lor.rhs7 ]
>   br label %land.end
> ```
> 
> Maybe do this in `simplifycfg` is fine? https://godbolt.org/z/6ohPTsGx3
And it seems correct
https://alive2.llvm.org/ce/z/o34sKD
